clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/metrics/UrlMetricsInstrumentation.groovy

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2018 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.docker.registry.metrics
+
+import com.netflix.spectator.api.Id
+import com.netflix.spectator.api.Registry
+import com.squareup.okhttp.Response
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+
+import static java.lang.String.format
+
+@Component
+class UrlMetricsInstrumentation {
+
+  private Logger log = LoggerFactory.getLogger(UrlMetricsInstrumentation)
+
+  private final Registry registry
+
+  private final Id timingId
+  private final Id counterId
+
+  @Autowired
+  UrlMetricsInstrumentation(Registry registry) {
+    this.registry = registry
+    timingId = registry.createId('UrlRequestTime').withTag('className', UrlMetricsInstrumentation.simpleName)
+    counterId = registry.createId('UrlCount').withTag('className', UrlMetricsInstrumentation.simpleName)
+  }
+
+  Id getTimingId() {
+    return timingId
+  }
+
+  Id getCounterId() {
+    return counterId
+  }
+
+  void interceptorCompleted(Response response, Instant start, Instant end) {
+    Duration duration = Duration.between(start, end)
+    URL url = response.request().url()
+    registry.timer(timingId.withTag('host', url.getHost()).withTag('path', url.getPath())).record(duration.toMillis(), TimeUnit.MILLISECONDS)
+    registry.counter(counterId.withTag('host', url.getHost()).withTag('path', url.getPath()).withTag('successful', response.isSuccessful())).increment()
+    log.info format("Received %d response for %s in %dms", response.code(), url, duration.toMillis())
```

Comment:
Do you need to log these given you just instrumented it ?

Reply:
Good question!

We (I used to be in the team that manages the Metrics, Monitoring and Logging stores, back-ends and infrastructure) always tell developers *"If it's not an error; don't log it, but instrument it instead"*, and here I find myself logging something that's not an error.

To show why, here's the full logging output for a single `DockerRegistryImageCachingAgent` execution:

```
2018-07-26 12:24:53.790  INFO 52852 --- [ecutionAction-1] c.n.s.c.d.r.m.UrlMetricsInstrumentation  : Received 200 response for https://eu.gcr.io/v2/bolcom-stg-spinnaker-6b3/provisioning-image/tags/list in 252ms
2018-07-26 12:24:53.791  INFO 52852 --- [ecutionAction-1] .d.r.p.a.DockerRegistryImageCachingAgent : Describing items in my-docker-registry/DockerRegistryImageCachingAgent[1/1]
2018-07-26 12:24:53.791  INFO 52852 --- [ecutionAction-1] .d.r.p.a.DockerRegistryImageCachingAgent : Caching 1 tagged images in my-docker-registry/DockerRegistryImageCachingAgent[1/1]
2018-07-26 12:24:53.791  INFO 52852 --- [ecutionAction-1] .d.r.p.a.DockerRegistryImageCachingAgent : Caching 1 image ids in my-docker-registry/DockerRegistryImageCachingAgent[1/1]
2018-07-26 12:24:53.794  INFO 52852 --- [ecutionAction-1] c.n.s.c.cache.LoggingInstrumentation     : com.netflix.spinnaker.clouddriver.docker.registry.provider.DockerRegistryProvider:my-docker-registry/DockerRegistryImageCachingAgent[1/1] completed in 0.256s
```

5 lines of logs, of which (in my opinion) 1 (Describing items) is useless, 3 are moderately informative (they contain a single variable bit of information), and the one this PR adds is the most informative (since it adds 3 pieces of information, and provides valuable context to the total execution time, since apparently the HTTP call takes up almost all of the time taken).

I added it so we could get a quick glance in Kibana as well, giving more context to the logs around them. That said, I totally understand why we'd want to keep the logging down to a minimum and could live without this line. (Although I think other, less informative, log events should be removed instead)

Let me know if you want it to stay or go, and I'll adjust it accordingly.

Reply:
Use your discretion. As I noted above, if you are logging calls then you dont need to instrument the /{path} calls for now. So you should keep at least some of this.